Loading spinner

krhobeck opened the pull request:
Added loading spinner to both the dashboard and parking map

Updated sidebar spinner for the dashboard map

krhobeck committed:
Updated existing spinner module, added to multiple places, changed filter tab loading

krhobeck committed:
Fixed linting

krhobeck committed:
Fixed the cloudflare type issue

krhobeck committed:
Fixed bug where zooming in causes a loading spinner on the dashboard

sonalranjit commented on bikespace_frontend/src/components/spinner/spinner.module.scss:
```diff
@@ -0,0 +1,46 @@
+//  100% {
+//    transform: rotate(360deg);
+//  }
+//}
```
Maybe we can delete dead code, the changes will be tracked in git if we need to revert it back later

sonalranjit commented on bikespace_frontend/src/components/dashboard/report-summary/report-summary.module.scss:
```diff
@@ -21,13 +21,13 @@
+//  font-size: 2rem;
+//  font-weight: 600;
+//  // text-align: center;
+//}
```
Same here, perhaps we can just delete the dead code

sonalranjit commented:
Thanks @krhobeck ! it's great addition with the spinner. I'll let @tallcoleman add his review, he's got a bit more frontend expertise than me

krhobeck commented:
No problem, I'll delete that dead code when I can

krhobeck committed:
Removed commented-out CSS from spinner and report-summary modules.

sonalranjit commented:
@krhobeck I notice a bit of a lag when zooming in and out, I guessing it's because the markers now re-render once the tiles are loaded. 
![screen-recording-1](https://github.com/user-attachments/assets/807e1089-ca6e-487e-9fb9-d422c35960dc)

Compared to what it is currently where the markers stay rendered even when zooming in and out of layers.
![screen-recording-2](https://github.com/user-attachments/assets/99102a60-2959-49b9-a977-3126a3d599d9)

If possible could we look to maintain the current behavior

krhobeck commented:
Ah for sure, let me look into fixing this

tallcoleman commented:
Thanks for doing this @krhobeck ! I'll try to take a look when I have time this week

tallcoleman committed:
Merge branch 'main' into loading-state-spinner

tallcoleman committed:
Move spinner into shared-ui

tallcoleman commented on bikespace_frontend/src/components/dashboard/map/Map.tsx:
```diff
@@ -1,23 +1,26 @@
 import {SubmissionApiPayload} from '@/interfaces/Submission';
 import {useSidebarTab} from '@/states/url-params';
 
+import {Spinner} from '@/components/shared-ui/spinner';
```
@krhobeck in #273 I added a components folder named shared-ui and have put the spinner there too. (After updating from main).

Also updated some import stuff:

- For a lot of component imports, we use an index.ts and import from the folder instead of from the component .tsx file. I guess we do this in case there are multiple related things you would want to export from the folder? But it keeps the import paths simpler too.
- Do you know if there are any tools / best practices for ordering imports? Based on reading the code that other frontend BikeSpace devs have written, I've internalized something like the following, but I have no idea how it relates to best practices:
  - framework imports (e.g. React, next)
  - third party dependencies (e.g. plotly, maplibre)
  - internal stuff in other folders (usually @-referenced, sometimes broken out by hooks, components, etc.)
  - internal stuff in the same folder (e.g. related components or ts files; usually referenced via relative path)
  - styles (all together regardless of whether they're in the same folder?)
  - assets (e.g. images, svgs)

krhobeck replied:
I'm not sure about best practices but the one you listed here sounds good to me, I generally follow the pattern of doing

1. Framework imports
2. Third party dependencies
3. 'Large' imports like large components or state stores or API types
4. 'Small' imports like reusables or any separate functions or types [this could also be small components, I usually put these first]
5. Styles & Assets

(I have a small exception for API types i usually put these closer to the top)

I guess I usually don't order based on folder but based on importance/size. For import sorting we could enforce a policy using eslint for this using a library like https://github.com/lydell/eslint-plugin-simple-import-sort and have regexes for this.

tallcoleman replied:
Thanks, I'll look into that plugin. Have started a discussion thread for this here: https://github.com/bikespace/bikespace/discussions/292

tallcoleman commented on bikespace_frontend/src/components/dashboard/map/Map.tsx:
```diff
@@ -1,23 +1,26 @@
+import 'leaflet-defaulticon-compatibility/dist/leaflet-defaulticon-compatibility.css';
+
 import styles from './map.module.scss';
+import loader from '@/styles/shared/loader.module.scss';
```
This import appears to be unused

tallcoleman commented on bikespace_frontend/src/components/dashboard/map/Map.tsx:
```diff
@@ -32,6 +35,8 @@ function Map({submissions}: MapProps) {
 
   const [doneLoading, setDoneLoading] = useState(false);
+  const [tilesLoaded, setTilesLoaded] = useState(false);
+  const isMapLoading = !doneLoading;
```
it looks like every use of `doneLoading` (including in `MapMarker` passed via prop) is inverting the value, so you could just change the main hook here to `[isMapLoading, setIsMapLoading] = useState(true)`, update the downstream values to un-invert them, and remove this const.

tallcoleman replied:
Resolved in https://github.com/bikespace/bikespace/pull/286/commits/8857a55323cd9a5d325a69029a34ba72fd4f3f45

tallcoleman commented on bikespace_frontend/src/components/dashboard/map/Map.tsx:
```diff
@@ -44,6 +49,9 @@ function Map({submissions}: MapProps) {
     if (!mapRef.current) return;
     mapRef.current.invalidateSize();
   }, [isSidebarOpen, currentSidebarTab]);
+  useEffect(() => {
```
Want to check my understanding of what's happening here - let me know if this sounds right?

- default state is "map is loading"
- if submissions.length changes, flips state to "map is loading"
- only once the final marker is drawn (see function in MapMarker ref on line 86) then state flipped to "map done loading"

I don't know React well enough - is there any chance of a race condition here? Are we sure that the marker rendering is always going to happen at least one cycle after submissions.length changes?

krhobeck replied:
I won't lie and I say I know react that well either. I think this version of the code could have that risk where the last marker mounts we call `setDoneLoading` to true but then it gets stomped by the `useEffect` and we have a spinner that is stuck on the screen when all the markers are showing. (This is real and could happen if `submissions.length` changed from 10 -> 12 lets say, see below).

I looked into this a bit and `useEffect` runs **after** the commit phase while ref callbacks run for each mount/unmount **during** the commit phase [here](https://stackoverflow.com/questions/75088157/are-refs-guaranteed-to-be-set-when-useeffect-runs). This means if there are no submissions in the array then the loader will literally run forever (which is questionable design).

tallcoleman replied:
> if there are no submissions in the array then the loader will literally run forever

I guess in this situation we have other problems anyway, so I won't worry about it for this PR, especially since in our current set-up, the data loads once and either succeeds or fails. We do have an issue to look into [changing up the data loading](https://github.com/bikespace/bikespace/issues/223) so I added a note there in case there's an issue with the data loading/updating multiple times while the user is using the page.

tallcoleman commented on bikespace_frontend/src/components/dashboard/map/Map.tsx:
```diff
@@ -59,27 +67,40 @@ function Map({submissions}: MapProps) {
         attribution='&copy; Maps <a href="https://www.thunderforest.com/">Thunderforest</a>, &copy; Data <a href="https://www.openstreetmap.org/copyright">OpenStreetMap contributors</a>'
         url="https://tile.thunderforest.com/atlas/{z}/{x}/{y}.png?apikey=66ccf6226ef54ef38a6b97fe0b0e5d2e"
         maxZoom={20}
+        eventHandlers={{
```
This is neat! I should look up whether it's possible to do something similar in MapLibre

tallcoleman commented on bikespace_frontend/src/components/dashboard/map/Map.tsx:
```diff
@@ -59,27 +67,40 @@ function Map({submissions}: MapProps) {
-          );
-        })}
-      </MarkerClusterGroup>
+      {tilesLoaded && ( // Tiles should be loaded before rendering markers
```
What happens if the markers render before the tiles?

tallcoleman replied:
Also, this state check inside the `MapContainer` is what is causing the markers to re-render on every zoom. (The issue that Sonal noted).

Maybe return a separate div with the spinner instead of the `MapContainer` while it's in the initial loading state? That would also ensure the spinner shows before the Map renders - currently, because the spinner is part of `MapContainer`, the spinner doesn't show up until Leaflet starts rendering at least the bare bones of the map.

tallcoleman replied:
Fixed in https://github.com/bikespace/bikespace/pull/286/commits/8857a55323cd9a5d325a69029a34ba72fd4f3f45 (or possibly earlier?)

tallcoleman commented on bikespace_frontend/src/components/dashboard/sidebar-content/_SidebarContentInsights.tsx:
```diff
@@ -1,13 +1,15 @@
-import React from 'react';
+import React, {Suspense, useCallback, useState} from 'react';
```
Suspense appears to be un-used

tallcoleman commented on bikespace_frontend/src/components/dashboard/sidebar-content/_SidebarContentInsights.tsx:
```diff
@@ -32,12 +34,24 @@ const DataDurationByTodChart = dynamic<ChartProps>(
 export function SidebarContentInsights() {
+  const TOTAL = 3;
+  const [readyCount, setReadyCount] = useState(0);
+  const markReady = useCallback(
```
is useCallback necessary here?

krhobeck replied:
It's actually not if you put a guard on each child, I'll update this

tallcoleman replied:
Resolved in https://github.com/bikespace/bikespace/pull/286/commits/865d3899045fc1db871d2ce2cbf4cc9754934238

tallcoleman commented on bikespace_frontend/src/components/dashboard/sidebar-content/_SidebarContentInsights.tsx:
```diff
@@ -32,12 +34,24 @@ const DataDurationByTodChart = dynamic<ChartProps>(
+        <Spinner overlay label="Loading insights..." style={{zIndex: 3000}} />
+      )}
+      <div style={{visibility: allReady ? 'visible' : 'hidden'}}>
+        <ReportSummary />
```
Since ReportSummary isn't one of the components that this spinner is waiting for, should it be moved above this spinner section?

krhobeck replied:
It should be, just added this

tallcoleman replied:
Thanks - I don't see any additional commits on the branch, were you planning to push them later?

tallcoleman commented on bikespace_frontend/src/components/parking-map/parking-map-page/parking-map-page.module.scss:
```diff
@@ -15,12 +15,20 @@
 .sideBarContainer {
   overflow-y: scroll;
   margin: 4px;
+  // hide scrollbar (firefox and edge)
```
My preference would be to add more margin left and right of the sidebar content instead of hiding the scroll bar, since the scroll bar could be considered an accessibility feature for some users. That way if there is a visible scroll bar, it doesn't overlap the content.

krhobeck replied:
Sure, good point!

krhobeck replied:
Do you mean padding instead of margin here? Adding a right padding will give the bar its own space on the side instead of merging with the sidebar content

tallcoleman replied:
Interesting... I didn't realize adding more margin moves the scrollbar too. 

Can you remind me which browser/OS combination you were trying to solve for here? Tried mac and windows x chrome & firefox x system settings on/off for persistent scrollbars and noticed the following:

- Persistent scrollbars seem to take up their own space instead of overlapping (which seems fine?)
- Auto-hiding scrollbars (I think is the default on mac and windows) overlap content while scrolling but hide again once the scroll action stops - there is an overlap potential here, but the current padding on the content boxes is usually good enough. Padding could be bumped up slightly if there's an OS/Browser combo we're aware of that has much thicker auto-hiding scrollbars.

krhobeck replied:
So I use brave + Linux mint, perhaps it has something to do with that, I'll push and take a look on my macbook to see if it looks weird. Here's a side by side comparison

<img width="414" height="848" alt="image" src="https://github.com/user-attachments/assets/9b6af52a-045c-4034-a581-6d2d420a9bd5" />

<img width="394" height="816" alt="image" src="https://github.com/user-attachments/assets/d04fd0be-4fd0-4c30-898c-8ef03f713747" />

The first image shows a padding-right by 6px, the second image is the padding-right removed with more margin set to 8px. But maybe the second image is the desired look? What are your thoughts?

tallcoleman replied:
The additional 6px padding you added in https://github.com/bikespace/bikespace/pull/286/commits/913c1fbcd7a013dc6c4c1125ca0a2be66dd7fb9e seems fine to me, if that resolves it for these OS/browser combinations you've tested. It's slightly asymmetrical, but it feels logical since the right side is the divider between the sidebar and the map.

Have left a separate comment that the added padding should be desktop only.

Thanks!

tallcoleman commented on bikespace_frontend/src/components/parking-map/parking-map-page/ParkingMapPage.tsx:
```diff
@@ -45,7 +46,7 @@ import type {
 import 'maplibre-gl/dist/maplibre-gl.css';
 import styles from './parking-map-page.module.scss';
-
+import loader from '@/styles/shared/loader.module.scss';
```
Import appears to not be used?

krhobeck replied:
Would you prefer if I added an eslint rule to catch these?

tallcoleman replied:
I think it would be just removing the current setting in `bikespace_frontend/.eslintrc.json:10-11`. 

I don't remember why we had that set, but I would guess someone was annoyed by having the linter fail the CI on the PR branch while they're in the middle of building out a feature. I guess people could just comment out an unused import until they use it later if they want to avoid the CI failure?

edit: maybe if it's just set to "warn" it won't fail the CI?

krhobeck replied:
I changed to set it to warn for now

tallcoleman commented on bikespace_frontend/src/components/parking-map/parking-map-page/ParkingMapPage.tsx:
```diff
@@ -220,6 +223,7 @@ export function ParkingMapPage() {
     if (process.env.NODE_ENV !== 'production') console.log('map loaded');
     addSprite();
     handleMouseHover();
+    console.log('done');
```
What does this do? There's a similar console.log above for testing, though it doesn't run in production.

tallcoleman commented on bikespace_frontend/src/components/parking-map/parking-map-page/ParkingMapPage.tsx:
```diff
@@ -307,6 +311,7 @@ export function ParkingMapPage() {
         }
         onLoad={handleOnLoad}
         onClick={handleLayerClick}
+        onIdle={() => setIsMapLoading(false)} // set to false when map is idle
```
Might be fine in this case, since this value currently only goes one way, but onIdle will fire every time the map stops doing something, including after pans and zooms that load additional vector tiles. May cause unneccessary re-renders?

tallcoleman replied:
Resolved in https://github.com/bikespace/bikespace/pull/286/commits/e3bcbde92c830a58614d3492ad0ebc04f2b403be

tallcoleman commented on bikespace_frontend/src/components/shared-ui/spinner/spinner.module.scss:
```diff
@@ -0,0 +1,25 @@
+@use './styles/variables.module.scss' as *;
```
I think we usually do `@import '@/styles/variables.module.scss';` Is there a reason you used `@use` in this case?

tallcoleman commented on bikespace_frontend/src/styles/shared/loader.module.scss:
This doesn't seem to be used in any of the files it's being imported by; should it be removed?

krhobeck replied:
This was a legacy loader design I deprecated, I can remove it

tallcoleman reviewed (changes requested):
Thanks @krhobeck this is really great! Will make sure that users don't think things are broken when they're just loading, especially if they're on mobile data or something slower. Have left some comments for you to take a look at. If you want help with fixing up anything to get it merge-ready, just let us know.

tallcoleman commented:
> @krhobeck I notice a bit of a lag when zooming in and out, I guessing it's because the markers now re-render once the tiles are loaded.

@krhobeck @sonalranjit have noted what I think is the cause of the issue here: https://github.com/bikespace/bikespace/pull/286/files#r2386171017

krhobeck committed:
Improved onReady handling for charts and report summary, added eslint-plugin-unused-imports, enabled stricter TS options, and updated CSS for scrollbar padding.

krhobeck committed:
Refactored map loading state: improved rendering logic, removed unused imports, and streamlined marker initialization.

krhobeck committed:
Refactored map loading logic

krhobeck committed:
Adjusted ESLint unused-imports rules

krhobeck committed:
Removed ts rule for unused variables

krhobeck commented:
Hey @tallcoleman I made some new changes, some of the old implementations i've scrapped completely, I decided to just take a simpler approach and keep an initial loading on the page.

tallcoleman commented on bikespace_frontend/src/components/parking-map/parking-map-page/parking-map-page.module.scss:
```diff
@@ -15,14 +15,7 @@
-    width: 0;
-    height: 0;
-  }
+  padding-right: 6px;
```
Can you please set this so it only applies in the desktop layout (i.e. when the viewport width is > $wrapper-full-width)?

krhobeck replied:
Updated this in the latest commit, let me know if that was the implementation you were looking for.

tallcoleman commented on bikespace_frontend/src/components/shared-ui/spinner/spinner.module.scss:
```diff
@@ -1,4 +1,4 @@
-@use './styles/variables.module.scss' as *;
+@import '@/styles/variables.module.scss';
```
I looked this up and realized that you were totally right to prefer `@use` over `@import` - I had no idea that `@import` had been deprecated.

Will open a new issue for converting the usage throughout the codebase since we'll need to think about whether we want to just `import as *` like you've done here or convert to the more default namespaced imports from `@use`.

krhobeck replied:
Sure no problem

tallcoleman committed:
Spacing fix for insights tab content boxes

The default content spacing only applies to the direct children of the holding div (implemented with flexbox gap). Added styling to ensure the same spacing is also applied to the children of the visibility control div.

tallcoleman commented on bikespace_frontend/src/components/parking-map/parking-map-page/ParkingMapPage.tsx:
```diff
@@ -201,6 +202,7 @@ export function ParkingMapPage() {
 
   function addSprite() {
     mapRef.current!.addSprite('parking', parkingSpritePath);
+    console.log('loading...');
```
Can this be removed or does it need to stay in (e.g. for testing purposes?)

krhobeck replied:
This has been removed in the latest commit

tallcoleman commented on bikespace_frontend/src/components/parking-map/parking-map-page/ParkingMapPage.tsx:
```diff
@@ -221,6 +222,10 @@ export function ParkingMapPage() {
+
+    // after styles load
+    const map: MapLibreMap = mapRef.current!.getMap(); // maplibre-gl map instance
+    map?.once('idle', () => setIsMapLoading(false)); // wait for styles to load, then set loading to false
```
I like this. Very elegant!

tallcoleman commented on bikespace_frontend/src/components/dashboard/sidebar-content/_SidebarContent.module.scss:
```diff
@@ -58,6 +58,12 @@ h2.cardHeading {
+.InnerMultipleSidebarContent {
+  display: flex;
+  flex-direction: column;
+  gap: 6px;
```
I would recommend you leave a comment above this gap value to explain the situation it addresses and for which browser/OS combos

krhobeck replied:
Good catch, this slipped my mind

tallcoleman reviewed (changes requested):
Thanks @krhobeck this is looking really good now.

Have gone through and resolved the old comments I left that are no longer relevant. Please take a look at the following that still need attention, in addition to the ones in this review:

- https://github.com/bikespace/bikespace/pull/286#discussion_r2386146675
- https://github.com/bikespace/bikespace/pull/286#discussion_r2386062881 (no need to answer the race condition question if it's a time suck)

I also pushed a small CSS fix in https://github.com/bikespace/bikespace/pull/286/commits/40a5f77b15f23091754e9f2181732d160890cf90 ; feel free to change/adjust it if you think a different approach would be better.

krhobeck committed:
Fixed sidebar padding issue for Linux Chrome browsers and removed unused console log in map component

krhobeck committed:
Aligned sidebar spacing for consistent rendering on Linux Chrome browsers.

tallcoleman reviewed (approved):
Thanks again @krhobeck for all your hard work on this! 

I think you should be able to squash merge now - just make sure to clean up the description on the merge commit.
